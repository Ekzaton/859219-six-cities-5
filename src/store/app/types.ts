import {ActionType} from "../const";
import {FilteringType, SortingType} from "../../const";

export type AppState = {
  activeOfferID: number | null,
  currentFiltering: FilteringType,
  currentSorting: SortingType
};

type getActiveOfferIDAction = {
  type: ActionType.GET_ACTIVE_OFFER_ID,
  payload: number | null
}

type getCurrentFilteringAction = {
  type: ActionType.GET_CURRENT_FILTERING,
  payload: FilteringType
}

type getCurrentSortingAction = {
  type: ActionType.GET_CURRENT_SORTING,
  payload: SortingType
}

export type AppAction = getActiveOfferIDAction | getCurrentFilteringAction | getCurrentSortingAction;
