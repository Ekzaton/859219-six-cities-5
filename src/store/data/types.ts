import {ActionType} from "../const";
import {Offer} from "../../components/types";

export type DataState = {
  offers: Offer[]
};

type getOffersAction = {
  type: typeof ActionType.GET_OFFERS,
  payload: Offer[]
}

export type DataAction = getOffersAction;

