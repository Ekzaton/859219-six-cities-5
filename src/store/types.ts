import {ActionType} from "./const";
import {Offer} from "../components/types";

export type State = {
  currentFilter: string,
  currentSorting: string,
  offers: Offer[],
};

type getCurrentFilterAction = {
  type: typeof ActionType.GET_CURRENT_FILTER,
  payload: string
}

type getCurrentSorting = {
  type: typeof ActionType.GET_CURRENT_SORTING,
  payload: string
}

type getOffersAction = {
  type: typeof ActionType.GET_OFFERS,
  payload: Offer[]
}

export type Action = getCurrentFilterAction | getCurrentSorting | getOffersAction;
