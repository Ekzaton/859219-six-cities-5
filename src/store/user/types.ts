import {ActionType} from "../const";
import {AuthStatus} from "../../const";
import {User} from "../../types";

export type UserState = {
  authStatus: AuthStatus,
  isDataSending: boolean,
  isSendingError: boolean,
  userData: User
};

type getAuthStatusAction = {
  type: typeof ActionType.GET_AUTH_STATUS,
  payload: AuthStatus
}

type getUserDataAction = {
  type: typeof ActionType.GET_USER_DATA,
  payload: User
}

type redirectToRouteAction = {
  type: typeof ActionType.REDIRECT_TO_ROUTE,
  payload: string
}

type setDataSendingAction = {
  type: ActionType.SET_IS_DATA_SENDING,
  payload: boolean
}

type setSendingErrorAction = {
  type: ActionType.SET_IS_SENDING_ERROR,
  payload: boolean
}

export type UserAction = getAuthStatusAction | getUserDataAction | redirectToRouteAction | setDataSendingAction | setSendingErrorAction;
