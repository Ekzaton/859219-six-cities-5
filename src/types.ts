export type offerType = {
  bedrooms: number;
  city: cityType;
  description: string;
  goods: string[];
  host: userType;
  id: number;
  image: string[];
  isFavorite: boolean;
  isPremium: boolean;
  location: locationType;
  maxAdults: number;
  previewImage: string;
  price: number;
  rating: number;
  title: string;
  type: string;
}

export type cityType = {
  location: locationType;
  name: string;
}

export type userType = {
  avatarUrl: string;
  id: number;
  isPro: boolean;
  name: string;
}

export type locationType = {
  latitude: number;
  longitude: number;
  zoom: number;
}

export type reviewType = {
  comment: string;
  date: string;
  id: number;
  rating: number;
  user: userType;
}
