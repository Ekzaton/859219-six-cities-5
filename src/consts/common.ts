export enum AppRoute {
  MAIN = `/`,
  LOGIN = `/login/`,
  FAVORITES = `/favorites/`,
  OFFER_ID = `/offer/:id/`,
  OFFER = `/offer/`
}

export enum AuthStatus {
  AUTH = `auth`,
  NO_AUTH = `no-auth`
}

export enum FilteringType {
  PARIS = `Paris`,
  COLOGNE = `Cologne`,
  BRUSSELS = `Brussels`,
  AMSTERDAM = `Amsterdam`,
  HAMBURG = `Hamburg`,
  DUSSELDORF = `Dusseldorf`
}

export enum SortingType {
  POPULAR = `Popular`,
  PRICE_LOW_TO_HIGH = `Price: low to high`,
  PRICE_HIGH_TO_LOW = `Price: high to low`,
  TOP_RATED_FIRST = `Top rated first`
}
