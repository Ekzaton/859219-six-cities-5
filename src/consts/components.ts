export const OFFER_IMAGES_COUNT = 6;
export const OFFER_REVIEWS_COUNT = 10;
export const RATING_STARS_COUNT = 5;
export const REG_EXP_EMAIL = /^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/;
export const REG_EXP_PASSWORD = /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$/;
export const DELAY_DURATION = 250;

export enum BtnBigSize {
  WIDTH = 31,
  HEIGHT = 33
}

export enum BtnSize {
  WIDTH = 18,
  HEIGHT = 19
}

export enum BtnType {
  PLACE_CARD = `place-card`,
  PROPERTY = `property`
}

export enum CardImgSize {
  WIDTH = 260,
  HEIGHT = 200
}

export enum CardType {
  CITIES = `cities`,
  FAVORITES = `favorites`,
  NEAR_PLACES = `near-places`
}

export enum FavCardImgSize {
  WIDTH = 150,
  HEIGHT = 110
}

export enum HeaderType {
  MAIN = `main`,
  LOGIN = `login`
}

export enum MapIconSize {
  WIDTH = 30,
  HEIGHT = WIDTH
}

export enum MapIconUrl {
  PIN = `img/pin.svg`,
  PIN_ACTIVE = `img/pin-active.svg`
}

export enum MapLayer {
  TEMPLATE = `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
  ATTRIBUTION = `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
}

export enum MapType {
  CITIES = `cities`,
  PROPERTY = `property`
}

export enum ReviewDate {
  LOCALE = `en-US`,
  MONTH = `long`,
  YEAR = `numeric`
}

export enum ReviewLehgth {
  MIN = 50,
  MAX = 300
}
