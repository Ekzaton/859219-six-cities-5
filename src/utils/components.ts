import {ReviewDate, ReviewLehgth} from "../const";

export const getOfferType = (type: string): string => type[0].toUpperCase() + type.slice(1);

export const getRatingArray = (length: number): number[] => new Array(length).fill(``);

export const getRatingStars = (rating: number): string => `${rating * 20}%`;

export const getReviewDate = (date: string): string =>
  new Date(date).toLocaleString(ReviewDate.LOCALE, {month: ReviewDate.MONTH, year: ReviewDate.YEAR});

export const validateComment = (comment: string): boolean =>
  comment.length >= ReviewLehgth.MIN && comment.length <= ReviewLehgth.MAX;


