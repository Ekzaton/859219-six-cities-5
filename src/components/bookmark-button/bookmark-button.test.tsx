import React from 'react';
import {Provider} from 'react-redux';
import createMockStore from 'redux-mock-store';
import {render} from '@testing-library/react';

import {BtnType} from "../../consts/components";
import offersMock from "../../mock/offers.json";
import {Offer} from "../../types/common";

import BookmarkButton from "./bookmark-button";

describe(`Bookmark button renders correctly`, () => {
  const mockStore = createMockStore([]);
  const store = mockStore();

  it(`for place card (active)`, () => {
    const offer = offersMock[3] as Offer;

    const {container} = render(
        <Provider store={store}>
          <BookmarkButton
            offer={offer}
            type={BtnType.PLACE_CARD}
          />
        </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it(`for place card (inactive)`, () => {
    const offer = offersMock[3] as Offer;

    const {container} = render(
        <Provider store={store}>
          <BookmarkButton
            offer={offer}
            type={BtnType.PLACE_CARD}
          />
        </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it(`for property page (active)`, () => {
    const offer = offersMock[0] as Offer;

    const {container} = render(
        <Provider store={store}>
          <BookmarkButton
            offer={offer}
            type={BtnType.PROPERTY}
          />
        </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it(`for property page (inactive)`, () => {
    const offer = offersMock[0] as Offer;

    const {container} = render(
        <Provider store={store}>
          <BookmarkButton
            offer={offer}
            type={BtnType.PROPERTY}
          />
        </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});

