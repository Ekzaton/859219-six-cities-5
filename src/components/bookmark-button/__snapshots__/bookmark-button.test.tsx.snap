// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark button renders correctly for place card (active) 1`] = `
<div>
  <button
    class="place-card__bookmark-button place-card__bookmark-button--active button"
    type="button"
  >
    <svg
      class="place-card__bookmark-icon"
      height="19"
      width="18"
    >
      <use
        xlink:href="#icon-bookmark"
      />
    </svg>
    <span
      class="visually-hidden"
    >
      In bookmarks
    </span>
  </button>
</div>
`;

exports[`Bookmark button renders correctly for place card (inactive) 1`] = `
<div>
  <button
    class="place-card__bookmark-button place-card__bookmark-button--active button"
    type="button"
  >
    <svg
      class="place-card__bookmark-icon"
      height="19"
      width="18"
    >
      <use
        xlink:href="#icon-bookmark"
      />
    </svg>
    <span
      class="visually-hidden"
    >
      In bookmarks
    </span>
  </button>
</div>
`;

exports[`Bookmark button renders correctly for property page (active) 1`] = `
<div>
  <button
    class="property__bookmark-button false button"
    type="button"
  >
    <svg
      class="property__bookmark-icon"
      height="33"
      width="31"
    >
      <use
        xlink:href="#icon-bookmark"
      />
    </svg>
    <span
      class="visually-hidden"
    >
      To bookmarks
    </span>
  </button>
</div>
`;

exports[`Bookmark button renders correctly for property page (inactive) 1`] = `
<div>
  <button
    class="property__bookmark-button false button"
    type="button"
  >
    <svg
      class="property__bookmark-icon"
      height="33"
      width="31"
    >
      <use
        xlink:href="#icon-bookmark"
      />
    </svg>
    <span
      class="visually-hidden"
    >
      To bookmarks
    </span>
  </button>
</div>
`;
